File: src/go/internal/hotkey/hotkey.go
// Package hotkey provides hotkey management for AicodeX
package hotkey

// Hotkey represents a hotkey binding
type Hotkey struct {
	Key         string
	Modifiers   []string
	Description string
	Action      func()
}

// Matches checks if the given key and modifiers match this hotkey
func (h *Hotkey) Matches(key string, modifiers []string) bool {
	if h.Key != key {
		return false
	}
	if len(h.Modifiers) != len(modifiers) {
		return false
	}
	modMap := make(map[string]bool)
	for _, m := range modifiers {
		modMap[m] = true
	}
	for _, m := range h.Modifiers {
		if !modMap[m] {
			return false
		}
	}
	return true
}

// Trigger executes the hotkey action
func (h *Hotkey) Trigger() bool {
	if h.Action != nil {
		h.Action()
		return true
	}
	return false
}

// Manager manages hotkey bindings
type Manager struct {
	hotkeys []*Hotkey
	enabled bool
}

// NewManager creates a new hotkey manager
func NewManager() *Manager {
	return &Manager{
		hotkeys: make([]*Hotkey, 0),
		enabled: true,
	}
}

// IsEnabled returns whether the manager is enabled
func (m *Manager) IsEnabled() bool {
	return m.enabled
}

// Enable enables the hotkey manager
func (m *Manager) Enable() {
	m.enabled = true
}

// Disable disables the hotkey manager
func (m *Manager) Disable() {
	m.enabled = false
}

// Register registers a new hotkey
func (m *Manager) Register(h *Hotkey) bool {
	for _, existing := range m.hotkeys {
		if existing.Matches(h.Key, h.Modifiers) {
			return false
		}
	}
	m.hotkeys = append(m.hotkeys, h)
	return true
}

// Unregister removes a hotkey
func (m *Manager) Unregister(key string, modifiers []string) bool {
	for i, h := range m.hotkeys {
		if h.Matches(key, modifiers) {
			m.hotkeys = append(m.hotkeys[:i], m.hotkeys[i+1:]...)
			return true
		}
	}
	return false
}

// Process processes a key press
func (m *Manager) Process(key string, modifiers []string) bool {
	if !m.enabled {
		return false
	}
	for _, h := range m.hotkeys {
		if h.Matches(key, modifiers) {
			return h.Trigger()
		}
	}
	return false
}

// ListHotkeys returns all registered hotkeys
func (m *Manager) ListHotkeys() []*Hotkey {
	result := make([]*Hotkey, len(m.hotkeys))
	copy(result, m.hotkeys)
	return result
}

File: src/python/aicodex/hotkey.py
"""Hotkey management module for AicodeX."""

from dataclasses import dataclass, field
from typing import Callable


@dataclass
class Hotkey:
    """Represents a hotkey binding."""

    key: str
    modifiers: list[str] = field(default_factory=list)
    action: Callable[[], None] | None = None
    description: str = ""

    def matches(self, key: str, modifiers: list[str]) -> bool:
        """Check if the given key and modifiers match this hotkey."""
        return self.key == key and set(self.modifiers) == set(modifiers)

    def trigger(self) -> bool:
        """Trigger the hotkey action."""
        if self.action is not None:
            self.action()
            return True
        return False


class HotkeyManager:
    """Manager for hotkey bindings."""

    def __init__(self) -> None:
        """Initialize the hotkey manager."""
        self._hotkeys: list[Hotkey] = []
        self._enabled = True

    @property
    def is_enabled(self) -> bool:
        """Check if hotkey manager is enabled."""
        return self._enabled

    def enable(self) -> None:
        """Enable hotkey processing."""
        self._enabled = True

    def disable(self) -> None:
        """Disable hotkey processing."""
        self._enabled = False

    def register(self, hotkey: Hotkey) -> bool:
        """Register a new hotkey."""
        for existing in self._hotkeys:
            if existing.matches(hotkey.key, hotkey.modifiers):
                return False
        self._hotkeys.append(hotkey)
        return True

    def unregister(self, key: str, modifiers: list[str] | None = None) -> bool:
        """Unregister a hotkey."""
        modifiers = modifiers or []
        for i, hotkey in enumerate(self._hotkeys):
            if hotkey.matches(key, modifiers):
                self._hotkeys.pop(i)
                return True
        return False

    def process(self, key: str, modifiers: list[str] | None = None) -> bool:
        """Process a key press and trigger matching hotkey."""
        if not self._enabled:
            return False
        modifiers = modifiers or []
        for hotkey in self._hotkeys:
            if hotkey.matches(key, modifiers):
                return hotkey.trigger()
        return False

    def list_hotkeys(self) -> list[Hotkey]:
        """Return list of registered hotkeys."""
        return self._hotkeys.copy()
